infra/Pos/Slotting/MemState/Class.hs

```diff
@@ -8,24 +8,52 @@ import           Control.Monad.Trans (MonadTrans)
 import           Universum
 
 import           Pos.Core.Types      (EpochIndex, Timestamp)
-import           Pos.Slotting.Types  (SlottingData)
+import           Pos.Slotting.Types  (EpochSlottingData)
+
 
 -- | 'MonadSlotsData' provides access to data necessary for slotting to work.
+-- Both _current_ and _next_ epoch @SlottingData@ need to be present since it's impossible
+-- to create @SlottingData@ without them.
 class Monad m => MonadSlotsData m where
 
-    getSystemStart :: m Timestamp
+    getSystemStartM :: m Timestamp
+
+    getAllEpochIndicesM :: m [EpochIndex]
+
+    getCurrentEpochIndexM :: m EpochIndex
+
+    getNextEpochIndexM :: m EpochIndex
```

Comment:
Ok, so what you are saying is that a typeclass must contain only the minimal/crucial amount of information, and we should use other functions we to be derive everything else. Makes sense.

But then you might ask if that information that you have in a typeclass is crucial since you need other functions to redefine it :man_shrugging: 

I think that in this case the code best fits when a typeclass returns `getCurrentNextEpochIndex :: (CurrentEpochIndex, NextEpochIndex)` and we use standard functions like `fst` and `snd` to pick which one we mean.